### PR TITLE
Fix sort row loss and compound symbol iteration. refactor some links …

### PR DESCRIPTION
**Pull request by plorenz**
…and map code

### CODE REVIEW

File: storage/boltz/crud_test.go [hunk 1]
```diff
@@ -624,6 +625,150 @@ func (test *crudTest) testLinkCollection(_ *testing.T) {
 	test.NoError(err)
 }
```
Comment (plorenz) [inline storage/boltz/crud_test.go]: Add lots of simple tests for different levels of linked entities and combinations of single entity and set entity links

File: storage/boltz/indexes.go [hunk 2]
```diff
@@ -171,10 +171,10 @@ type SetReadIndex interface {
 }
```
Comment (plorenz) [inline storage/boltz/indexes.go]: Make interface methods public so we could add constraint types that are specific to a given store

File: storage/boltz/link_collection.go [hunk 3]
```diff
@@ -153,46 +157,118 @@ func (collection *linkCollectionImpl) GetLinks(tx *bbolt.Tx, id string) []string
...
 }
 
 func (collection *linkCollectionImpl) unlinkCursor(tx *bbolt.Tx, cursor *bbolt.Cursor, id []byte, associatedId []byte) error {
 	if err := cursor.Delete(); err != nil {
 		return err
 	}
-	return collection.unlinkOther(tx, id, associatedId)
+	return collection.otherField.RemoveLink(tx, associatedId, id)
+}
+
```
Comment (plorenz) [inline storage/boltz/link_collection.go]: Add LinkedSetSymbol, which extracts out some of the link/unlink code that was above and also adds ability to link/unlink  compound keys

File: storage/boltz/query.go [hunk 4]
```diff
@@ -48,11 +49,58 @@ type EntitySymbol interface {
 	Eval(tx *bbolt.Tx, rowId []byte) (FieldType, []byte)
 }
```
Comment (plorenz) [inline storage/boltz/query.go]: Simplify Map function API by using a context instead of a complicated return signature

File: storage/boltz/query_symbols.go [hunk 5]
```diff
@@ -340,6 +344,14 @@ type compositeEntitySetSymbol struct {
 	cursorLastF func(tx *bbolt.Tx, key []byte) (FieldType, []byte)
 }
```
Comment (plorenz) [inline storage/boltz/query_symbols.go]: This interface method was missing, which meant we didn't properly build chains when we had more than 1 linked entity in a symbol.

File: storage/boltz/query_test.go [hunk 6]
```diff
@@ -377,17 +377,36 @@ func (test *boltQueryTests) testBusinessEquals(*testing.T) {
 }
 
 func (test *boltQueryTests) testSortPage(*testing.T) {
-	//paging := &predicate.Paging{Offset: 0, Limit: 10}
-	//sorts := []predicate.SortField{{Field: "lastName", Direction: predicate.DESC}, {Field: "firstName", Direction: predicate.ASC}}
-	ids, count := test.query(`firstName in ["Alice", "Bob", "Cecilia", "David"] SORT BY lastName desc, firstName limit 10`)
```
Comment (plorenz) [inline storage/boltz/query_test.go]: Add test where we have duplication in sort parameters, to ensure we don't drop duplicate sort parameters

File: storage/boltz/store_query.go [hunk 7]
```diff
@@ -91,7 +91,9 @@ func (store *BaseStore) GetSymbol(name string) EntitySymbol {
 			if !ok || linkedEntitySymbol.getLinkedType() == nil {
 				return nil // Can only have composite symbols if it's linked
 			}
-			rest := linkedEntitySymbol.getLinkedType().GetSymbol(parts[1])
```
Comment (plorenz) [inline storage/boltz/store_query.go]: This was dropping symbols when we had more than one level of chained entities
